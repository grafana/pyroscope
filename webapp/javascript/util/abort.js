export function isAbortError(err) {
  if (!err) {
    return false;
  }

  // https://developer.mozilla.org/en-US/docs/Web/API/DOMException
  return err.name === 'AbortError' || err.code === 20;
}
