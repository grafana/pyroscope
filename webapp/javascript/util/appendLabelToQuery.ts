export function appendLabelToQuery(
  query: string,
  label: string,
  labelValue: string
) {
  const case1Regexp = new RegExp(`${label}=.+?(\\}|,)`);
  if (query.match(case1Regexp)) {
    return query.replace(case1Regexp, `${label}="${labelValue}"$1`);
  }
  if (query.indexOf('{}') !== -1) {
    return query.replace('}', `${label}="${labelValue}"}`);
  }
  if (query.indexOf('}') !== -1) {
    return query.replace('}', `, ${label}="${labelValue}"}`);
  }

  console.warn('TODO: handle this case');
  return query;
}
