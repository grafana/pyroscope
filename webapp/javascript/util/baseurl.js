// There's a copy of this function in welcome.html
//   TODO: maybe dedup somehow?
function basename() {
  const baseURLMetaTag = document.querySelector(
    'meta[name="pyroscope-base-url"]'
  );
  if (!baseURLMetaTag) {
    return null;
  }

  const baseURL = baseURLMetaTag.content;

  if (!baseURL) {
    return null;
  }
  const url = new URL(baseURL, window.location.href);
  return url.pathname;
}

export default basename;
