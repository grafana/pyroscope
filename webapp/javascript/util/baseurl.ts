// There's a copy of this function in welcome.html
//   TODO: maybe dedup somehow?
function basename() {
  const baseURLMetaTag = document.querySelector(
    'meta[name="pyroscope-base-url"]'
  ) as HTMLMetaElement;

  if (!baseURLMetaTag) {
    return undefined;
  }

  const baseURL = baseURLMetaTag.content;

  if (!baseURL) {
    return undefined;
  }
  const url = new URL(baseURL, window.location.href);
  return url.pathname;
}

export default basename;
