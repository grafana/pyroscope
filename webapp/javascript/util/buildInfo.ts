// eslint-disable-next-line import/no-relative-packages
import { version as jsVersion } from '../../../package.json';

export interface BuildInfo {
  goos: string;
  goarch: string;
  goVersion: string;
  version: string;
  time: string;
  gitSHA: string;
  gitDirty: string;
  useEmbeddedAssets: string;
  jsVersion: string;
}

export const buildInfo = function (): BuildInfo {
  // TODO: it may be possible that these fields are not populated
  const win = window as unknown as { buildInfo: BuildInfo };

  return {
    jsVersion,
    goos: win.buildInfo?.goos,
    goarch: win.buildInfo?.goarch,
    goVersion: win.buildInfo?.goVersion,
    version: win.buildInfo?.version,
    time: win.buildInfo?.time,
    gitSHA: win.buildInfo?.gitSHA,
    gitDirty: win.buildInfo?.gitDirty,
    useEmbeddedAssets: win.buildInfo?.useEmbeddedAssets,
  };
};

interface LatestVersionInfo {
  latest_version: string;
}

export const latestVersionInfo = function (): LatestVersionInfo {
  // TODO: it may be possible that these fields are not populated
  const win = window as unknown as { latestVersionInfo: LatestVersionInfo };

  return {
    latest_version: win.latestVersionInfo.latest_version,
  };
};
