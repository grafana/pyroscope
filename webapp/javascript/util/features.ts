// Features represents the list of UI experimental features
// This interfaces serves for when we introduce
// something more sofisticated than simply booleans
interface Features {
  enableExperimentalAdhocUI?: boolean;
}

function hasFeatures(
  window: unknown
): window is typeof window & { features: Features } {
  if (typeof window === 'object') {
    if (window.hasOwnProperty('features')) {
      return true;
    }
  }

  return false;
}

// Features
export const isExperimentalAdhocUIEnabled = hasFeatures(window)
  ? window.features.enableExperimentalAdhocUI
  : false;
