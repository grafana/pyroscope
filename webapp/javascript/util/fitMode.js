// Even though there are only 2 values
// Which could be represented in a boolean fashion
// Let's use an Enum in case we want to use new modes,
// for example a new heuristic based on the language
export const FitModes = {
  TAIL: "TAIL",
  HEAD: "HEAD",
}

const margin = 3;

/**
 * Returns a text and margin left used to write text into a canvas rectangle
 *
 * @param {FitModes} mode -
 * @param {number} charSize - Size in pixels of an individual character. Assumes it's a monospace font.
 * @param {number} rectWidth - Width in pixels of the rectangle
 * @param {string} fullText - The text that will be first tried.
 * @param {string} shortText - The text that willbe used when fullText can't fit. It's normally a substring of the original text.
 */
export function fitToCanvasRect({ mode, charSize, rectWidth, fullText, shortText }){
  switch (mode) {
    case FitModes.TAIL:
      // Case 1:
      // content fits rectangle width
      // | rectangle |
      // | text |
      if (charSize * fullText.length <= rectWidth) { // assume it's a monospaced font
        return {
          mode,
          text: fullText,
          marginLeft: margin,
        }
      }

      // assume it's a monospaced font
      // if not the case, use
      // ctx.measureText(shortName).width
      const shortTextWidth = charSize * shortText.length;

      // Case 2:
      // short text fits rectangle width
      // | rectangle |
      // | long_text_text |
      // | shorttext |
      if (shortTextWidth <= rectWidth) { // assume it's a monospaced font
        return {
          mode,
          text: shortText,
          marginLeft: margin,
        }
      }

      // Case 3:
      // short text is bigger than rectangle width
      // add a negative margin left
      // so that the 'tail' of the string is visible
      //     | rectangle |
      // | my_short_text |
      return {
        mode,
        text: shortText,
        marginLeft: -((shortTextWidth - rectWidth) + margin),
      }

    // Case 3:
    // Normal
    case FitModes.HEAD:
    default:
      return {
        mode,
        text: fullText,
        marginLeft: margin,
      }
  }
}


/**
 * Returns an inline style in React format
 * used to fit the content into a table cell
 * or an empty object if not applicable.
 * @param {FitModes} mode - The mode
 */
export function fitIntoTableCell(mode) {
  switch (mode) {
    case FitModes.TAIL:
      return {
       // prints from right to left
        direction: "rtl",
        overflow: "hidden",
        textOverflow: "ellipsis",
      };

    case FitModes.HEAD:
    default:
      return {
        overflow: "hidden",
        textOverflow: "ellipsis",
      };
  }
}
