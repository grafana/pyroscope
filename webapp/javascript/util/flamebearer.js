// ISC License

// Copyright (c) 2018, Mapbox

// Permission to use, copy, modify, and/or distribute this software for any purpose
// with or without fee is hereby granted, provided that the above copyright notice
// and this permission notice appear in all copies.

// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
// OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
// TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
// THIS SOFTWARE.

// This component is based on flamebearer project
//   https://github.com/mapbox/flamebearer

function deltaDiff(levels, start, step) {
  for (const level of levels) {
    let prev = 0;
    for (let i = start; i < level.length; i += step) {
      level[i] += prev;
      prev = level[i] + level[i + 1];
    }
  }
}

export function deltaDiffWrapper(format, levels) {
  if (format === "double") {
    deltaDiff(levels, 0, 7);
    deltaDiff(levels, 3, 7);

  } else {
    deltaDiff(levels, 0, 4);
  }
}

// format=single
//   j = 0: x start of bar
//   j = 1: width of bar
//   j = 3: position in the main index (jStep)
//
// format=double
//   j = 0,3: x start of bar =>     x = (level[0] + level[3]) / 2
//   j = 1,4: width of bar   => width = (level[4] + level[1]) / 2
//                           =>  diff = (level[4] - level[1]) / (level[1] + level[4])
//   j = 6  : position in the main index (jStep)

const formatSingle = {
  format: "single",
  jStep : 4,
  jName : 3,
  getBarOffset:    (level, j) => level[j],
  getBarTotal:     (level, j) => level[j + 1],
  getBarTotalDiff: (level, j) => 0,
  getBarSelf:      (level, j) => level[j + 2],
  getBarSelfDiff:  (level, j) => 0,
  getBarName:      (level, j) => level[j + 3],
}

const formatDouble = {
  format: "double",
  jStep : 7,
  jName : 6,
  getBarOffset:    (level, j) => (level[j]     + level[j + 3]),
  getBarTotal:     (level, j) => (level[j + 4] + level[j + 1]),
  getBarTotalLeft: (level, j) =>  level[j + 1],
  getBarTotalRght: (level, j) =>  level[j + 4],
  getBarTotalDiff: (level, j) => (level[j + 4] - level[j + 1]),
  getBarSelf:      (level, j) => (level[j + 5] + level[j + 2]),
  getBarSelfLeft:  (level, j) =>  level[j + 2],
  getBarSelfRght:  (level, j) =>  level[j + 5],
  getBarSelfDiff:  (level, j) => (level[j + 5] - level[j + 2]),
  getBarName:      (level, j) =>  level[j + 6],
}

export function parseFlamebearerFormat(format) {
  const isSingle = format !== "double";
  if (isSingle) return formatSingle;
  else return formatDouble;
}
