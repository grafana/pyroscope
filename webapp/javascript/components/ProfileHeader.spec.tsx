import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ProfileHeader, { TOOLBAR_MODE_WIDTH_THRESHOLD } from './ProfilerHeader';
import { FitModes } from '../util/fitMode';

function setWindowSize(s: 'small' | 'large') {
  const boundingClientRect = {
    x: 0,
    y: 0,
    bottom: 0,
    right: 0,
    toJSON: () => {},
    height: 0,
    top: 0,
    left: 0,
  };

  switch (s) {
    case 'large': {
      // https://github.com/jsdom/jsdom/issues/653#issuecomment-606323844
      window.HTMLElement.prototype.getBoundingClientRect = function () {
        return {
          ...boundingClientRect,
          width: TOOLBAR_MODE_WIDTH_THRESHOLD,
        };
      };
      break;
    }
    case 'small': {
      // https://github.com/jsdom/jsdom/issues/653#issuecomment-606323844
      window.HTMLElement.prototype.getBoundingClientRect = function () {
        return {
          ...boundingClientRect,
          width: TOOLBAR_MODE_WIDTH_THRESHOLD - 1,
        };
      };
      break;
    }

    default: {
      throw new Error('Wrong value');
    }
  }
}

describe('ProfileHeader', () => {
  it('shifts between visualization modes', () => {
    setWindowSize('large');

    const { asFragment, rerender } = render(
      <ProfileHeader
        view="both"
        handleSearchChange={() => {}}
        resetStyle={{}}
        reset={() => {}}
        updateFitMode={() => {}}
        fitMode={FitModes.HEAD}
        updateView={() => {}}
        updateViewDiff={() => {}}
      />
    );

    expect(screen.getByRole('toolbar')).toHaveAttribute('data-mode', 'large');
    expect(asFragment()).toMatchSnapshot();

    setWindowSize('small');

    rerender(
      <ProfileHeader
        view="both"
        handleSearchChange={() => {}}
        resetStyle={{}}
        reset={() => {}}
        updateFitMode={() => {}}
        fitMode={FitModes.HEAD}
        updateView={() => {}}
        updateViewDiff={() => {}}
      />
    );

    expect(screen.getByRole('toolbar')).toHaveAttribute('data-mode', 'small');
    expect(asFragment()).toMatchSnapshot();
  });

  describe('ViewSection', () => {
    const updateView = jest.fn();
    const component = (
      <ProfileHeader
        view="both"
        handleSearchChange={() => {}}
        resetStyle={{}}
        reset={() => {}}
        updateFitMode={() => {}}
        fitMode={FitModes.HEAD}
        updateView={updateView}
        updateViewDiff={() => {}}
      />
    );

    describe('large mode', () => {
      beforeEach(() => {
        setWindowSize('large');
        render(component);
      });

      it('changes to Table View', () => {
        screen.getByRole('button', { name: /Table/ }).click();
        expect(updateView).toHaveBeenCalledWith('table');
      });

      it('changes to Flamegraph view', () => {
        screen.getByRole('button', { name: /Flamegraph/ }).click();
        expect(updateView).toHaveBeenCalledWith('icicle');
      });

      it('changes to Both view', () => {
        screen.getByRole('button', { name: /Both/ }).click();
        expect(updateView).toHaveBeenCalledWith('both');
      });
    });

    describe('small mode', () => {
      beforeEach(() => {
        setWindowSize('small');
        render(component);
      });

      it('changes to Table view', () => {
        userEvent.selectOptions(
          screen.getByRole('combobox', { name: /view/ }),
          screen.getByRole('option', { name: /Table/ })
        );
        expect(updateView).toHaveBeenCalledWith('table');
      });

      it('changes to Flamegraph view', () => {
        userEvent.selectOptions(
          screen.getByRole('combobox', { name: /view/ }),
          screen.getByRole('option', { name: /Flamegraph/ })
        );
        expect(updateView).toHaveBeenCalledWith('icicle');
      });

      it('changes to Both view', () => {
        userEvent.selectOptions(
          screen.getByRole('combobox', { name: /view/ }),
          screen.getByRole('option', { name: /Both/ })
        );
        expect(updateView).toHaveBeenCalledWith('both');
      });
    });
  });
});
