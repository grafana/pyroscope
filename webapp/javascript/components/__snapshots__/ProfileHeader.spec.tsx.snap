// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileHeader shifts between visualization modes 1`] = `
<DocumentFragment>
  <div
    data-mode="large"
    role="toolbar"
  >
    <div
      class="navbar-2"
    >
      <input
        class="flamegraph-search"
        data-testid="flamegraph-search"
        name="flamegraph-search"
        placeholder="Search…"
      />
       
      <button
        class="btn"
        data-testid="reset-view"
        disabled=""
        id="reset"
        type="button"
      >
        Reset View
      </button>
      <select
        aria-label="fit-mode"
        class="fit-mode-select"
      >
        <option
          disabled=""
        >
          Prefer to fit
        </option>
        <option
          value="HEAD"
        >
          Head First
        </option>
        <option
          value="TAIL"
        >
          Tail First
        </option>
      </select>
      <div
        class="navbar-space-filler"
      />
      <div
        class="btn-group viz-switch"
      >
        <button
          class="btn"
          data-testid="btn-table-view"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-table fa-w-16 "
            data-icon="table"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64v-96h160v96zm0-160H64v-96h160v96zm224 160H288v-96h160v96zm0-160H288v-96h160v96z"
              fill="currentColor"
            />
          </svg>
            Table
        </button>
        <button
          class="btn active"
          data-testid="btn-both-view"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-columns fa-w-16 "
            data-icon="columns"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64V160h160v256zm224 0H288V160h160v256z"
              fill="currentColor"
            />
          </svg>
            Both
        </button>
        <button
          class="btn"
          data-testid="btn-flamegraph-view"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-icicles fa-w-16 "
            data-icon="icicles"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M511.4 37.9C515.1 18.2 500 0 480 0H32C10.6 0-4.8 20.7 1.4 41.2l87.1 273.4c2.5 7.2 12.7 7.2 15.1 0L140 190.5l44.2 187.3c1.9 8.3 13.7 8.3 15.6 0l46.5-196.9 34.1 133.4c2.3 7.6 13 7.6 15.3 0l45.8-172.5 66.7 363.8c1.7 8.6 14 8.6 15.7 0l87.5-467.7z"
              fill="currentColor"
            />
          </svg>
            Flamegraph
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProfileHeader shifts between visualization modes 2`] = `
<DocumentFragment>
  <div
    data-mode="small"
    role="toolbar"
  >
    <div
      class="navbar-2"
    >
      <input
        class="flamegraph-search"
        data-testid="flamegraph-search"
        name="flamegraph-search"
        placeholder="Search…"
      />
       
      <button
        class="btn"
        data-testid="reset-view"
        disabled=""
        id="reset"
        type="button"
      >
        Reset View
      </button>
      <select
        aria-label="fit-mode"
        class="fit-mode-select"
      >
        <option
          disabled=""
        >
          Prefer to fit
        </option>
        <option
          value="HEAD"
        >
          Head First
        </option>
        <option
          value="TAIL"
        >
          Tail First
        </option>
      </select>
      <div
        class="navbar-space-filler"
      />
      <div
        class="btn-group viz-switch"
      >
        <select
          aria-label="view"
        >
          <option
            value="table"
          >
            Table
          </option>
          <option
            value="both"
          >
            Both
          </option>
          <option
            value="icicle"
          >
            Flamegraph
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
